SHM provider compilation warning

Got the following warning from SHM provider. Looks like a bug to me:

```
  CC       prov/shm/src/src_libfabric_la-smr_rma.lo
prov/shm/src/smr_progress.c: In function 'smr_progress_resp_entry':
prov/shm/src/smr_progress.c:123:34: warning: comparison of unsigned expression < 0 is always false [-Wtype-limits]
      *err = (pending->bytes_done < 0)
                                  ^

prov/shm/src/smr_progress.c:151:32: warning: comparison of unsigned expression < 0 is always false [-Wtype-limits]
    *err = (pending->bytes_done < 0)
                                ^
prov/shm/src/smr_progress.c: In function 'smr_progress_inline':
prov/shm/src/smr_progress.c:226:22: warning: comparison of unsigned expression < 0 is always false [-Wtype-limits]
   return (*total_len < 0) ? ((int) *total_len) : -FI_ETRUNC;
                      ^
prov/shm/src/smr_progress.c: In function 'smr_progress_inject':
prov/shm/src/smr_progress.c:259:22: warning: comparison of unsigned expression < 0 is always false [-Wtype-limits]
   return (*total_len < 0) ? ((int) *total_len) : -FI_ETRUNC;
                      ^
prov/shm/src/smr_progress.c: In function 'smr_mmap_peer_copy':
prov/shm/src/smr_progress.c:334:22: warning: comparison of unsigned expression < 0 is always false [-Wtype-limits]
    ret = (*total_len < 0)
                      ^
prov/shm/src/smr_progress.c:345:22: warning: comparison of unsigned expression < 0 is always false [-Wtype-limits]
    ret = (*total_len < 0)
                      ^
prov/shm/src/smr_progress.c: In function 'smr_progress_ipc':
prov/shm/src/smr_progress.c:487:21: warning: comparison of unsigned expression < 0 is always false [-Wtype-limits]
   ret = (*total_len < 0) ? ((int) *total_len) : -FI_ETRUNC;
```

--- a/prov/shm/src/smr_progress.c
+++ b/prov/shm/src/smr_progress.c
@@ -72,6 +72,7 @@ static int smr_progress_resp_entry(struct smr_ep *ep, struct smr_resp *resp,
 	struct smr_inject_buf *tx_buf = NULL;
 	struct smr_sar_msg *sar_msg = NULL;
 	uint8_t *src;
+	ssize_t hmem_copy_ret;
 
 	peer_smr = smr_peer_region(ep->region, pending->peer_id);
 
@@ -110,19 +111,24 @@ static int smr_progress_resp_entry(struct smr_ep *ep, struct smr_resp *resp,
 			break;
 		if (pending->cmd.msg.hdr.op == ofi_op_read_req) {
 			if (!*err) {
-				pending->bytes_done =
+				hmem_copy_ret =
 					ofi_copy_to_hmem_iov(pending->iface,
 							     pending->device,
 							     pending->iov,
 							     pending->iov_count,
 							     0, pending->map_ptr,
 							     pending->cmd.msg.hdr.size);
-				if (pending->bytes_done != pending->cmd.msg.hdr.size) {
+				if (hmem_copy_ret < 0) {
+					FI_WARN(&smr_prov, FI_LOG_EP_CTRL,
+						"Copy from mmapped file failed with code %d\n",
+						(int)(-hmem_copy_ret));
+					*err = hmem_copy_ret;
+				} else if (hmem_copy_ret != pending->cmd.msg.hdr.size) {
 					FI_WARN(&smr_prov, FI_LOG_EP_CTRL,
 						"Incomplete copy from mmapped file\n");
-					*err = (pending->bytes_done < 0)
-						? ((int) pending->bytes_done)
-						: -FI_ETRUNC;
+					*err = -FI_ETRUNC;
+				} else {
+					pending->bytes_done = (size_t) hmem_copy_ret;
 				}
 			}
 			munmap(pending->map_ptr, pending->cmd.msg.hdr.size);
@@ -141,15 +147,21 @@ static int smr_progress_resp_entry(struct smr_ep *ep, struct smr_resp *resp,
 
 		src = pending->cmd.msg.hdr.op == ofi_op_atomic_compare ?
 		      tx_buf->buf : tx_buf->data;
-		pending->bytes_done = ofi_copy_to_hmem_iov(pending->iface, pending->device,
-							   pending->iov, pending->iov_count,
-							   0, src, pending->cmd.msg.hdr.size);
+		hmem_copy_ret  = ofi_copy_to_hmem_iov(pending->iface, pending->device,
+						      pending->iov, pending->iov_count,
+						      0, src, pending->cmd.msg.hdr.size);
 
-		if (pending->bytes_done != pending->cmd.msg.hdr.size) {
+		if (hmem_copy_ret < 0) {
+			FI_WARN(&smr_prov, FI_LOG_EP_CTRL,
+				"RMA read/fetch failed with code %d\n",
+				(int)(-hmem_copy_ret));
+			*err = hmem_copy_ret;
+		} else if (hmem_copy_ret != pending->cmd.msg.hdr.size) {
 			FI_WARN(&smr_prov, FI_LOG_EP_CTRL,
 				"Incomplete rma read/fetch buffer copied\n");
-			*err = (pending->bytes_done < 0)
-				? ((int) pending->bytes_done) : -FI_ETRUNC;
+			*err = -FI_ETRUNC;
+		} else {
+			pending->bytes_done = (size_t) hmem_copy_ret;
 		}
 		break;
 	default:
@@ -218,13 +230,23 @@ static int smr_progress_inline(struct smr_cmd *cmd, enum fi_hmem_iface iface,
 			       uint64_t device, struct iovec *iov,
 			       size_t iov_count, size_t *total_len)
 {
-	*total_len = ofi_copy_to_hmem_iov(iface, device, iov, iov_count, 0,
-					  cmd->msg.data.msg, cmd->msg.hdr.size);
-	if (*total_len != cmd->msg.hdr.size) {
+	ssize_t hmem_copy_ret;
+
+	hmem_copy_ret = ofi_copy_to_hmem_iov(iface, device, iov, iov_count, 0,
+					     cmd->msg.data.msg, cmd->msg.hdr.size);
+	if (hmem_copy_ret < 0) {
 		FI_WARN(&smr_prov, FI_LOG_EP_CTRL,
-			"recv truncated");
-		return (*total_len < 0) ? ((int) *total_len) : -FI_ETRUNC;
+			"inline recv failed with code %d\n",
+			(int)(-hmem_copy_ret));
+		return hmem_copy_ret;
+	} else if (hmem_copy_ret != cmd->msg.hdr.size) {
+		FI_WARN(&smr_prov, FI_LOG_EP_CTRL,
+			"inline recv truncated\n");
+		return -FI_ETRUNC;
 	}
+
+	*total_len = hmem_copy_ret;
+
 	return FI_SUCCESS;
 }
 
@@ -235,6 +257,7 @@ static int smr_progress_inject(struct smr_cmd *cmd, enum fi_hmem_iface iface,
 {
 	struct smr_inject_buf *tx_buf;
 	size_t inj_offset;
+	ssize_t hmem_copy_ret;
 
 	inj_offset = (size_t) cmd->msg.hdr.src_data;
 	tx_buf = smr_get_ptr(ep->region, inj_offset);
@@ -245,19 +268,29 @@ static int smr_progress_inject(struct smr_cmd *cmd, enum fi_hmem_iface iface,
 	}
 
 	if (cmd->msg.hdr.op == ofi_op_read_req) {
-		*total_len = ofi_copy_from_hmem_iov(tx_buf->data, cmd->msg.hdr.size,
-						    iface, device, iov, iov_count, 0);
+		hmem_copy_ret = ofi_copy_from_hmem_iov(tx_buf->data,
+						       cmd->msg.hdr.size,
+						       iface, device, iov,
+						       iov_count, 0);
 	} else {
-		*total_len = ofi_copy_to_hmem_iov(iface, device, iov, iov_count, 0,
-						  tx_buf->data, cmd->msg.hdr.size);
+		hmem_copy_ret = ofi_copy_to_hmem_iov(iface, device, iov,
+						     iov_count, 0, tx_buf->data,
+						     cmd->msg.hdr.size);
 		smr_freestack_push(smr_inject_pool(ep->region), tx_buf);
 	}
 
-	if (*total_len != cmd->msg.hdr.size) {
+	if (hmem_copy_ret < 0) {
 		FI_WARN(&smr_prov, FI_LOG_EP_CTRL,
-			"recv truncated");
-		return (*total_len < 0) ? ((int) *total_len) : -FI_ETRUNC;
+			"inject recv failed with code %d\n",
+			(int)(-hmem_copy_ret));
+		return hmem_copy_ret;
+	} else if (hmem_copy_ret != cmd->msg.hdr.size) {
+		FI_WARN(&smr_prov, FI_LOG_EP_CTRL,
+			"inject recv truncated\n");
+		return -FI_ETRUNC;
 	}
+
+	*total_len = hmem_copy_ret;
 
 	return FI_SUCCESS;
 }
@@ -301,6 +334,7 @@ static int smr_mmap_peer_copy(struct smr_ep *ep, struct smr_cmd *cmd,
 	void *mapped_ptr;
 	int fd, num;
 	int ret = 0;
+	ssize_t hmem_copy_ret;
 
 	num = smr_mmap_name(shm_name,
 			ep->region->map->peers[cmd->msg.hdr.id].peer.name,
@@ -325,30 +359,28 @@ static int smr_mmap_peer_copy(struct smr_ep *ep, struct smr_cmd *cmd,
 	}
 
 	if (cmd->msg.hdr.op == ofi_op_read_req) {
-		*total_len = ofi_copy_from_hmem_iov(mapped_ptr,
+		hmem_copy_ret = ofi_copy_from_hmem_iov(mapped_ptr,
 						    cmd->msg.hdr.size, iface,
 						    device, iov, iov_count, 0);
-		if (*total_len != cmd->msg.hdr.size) {
-			FI_WARN(&smr_prov, FI_LOG_EP_CTRL,
-				"mmap iov copy in error\n");
-			ret = (*total_len < 0)
-				? ((int) *total_len) : -FI_ETRUNC;
-			goto munmap;
-		}
 	} else {
-		*total_len = ofi_copy_to_hmem_iov(iface, device, iov,
+		hmem_copy_ret = ofi_copy_to_hmem_iov(iface, device, iov,
 						  iov_count, 0, mapped_ptr,
 						  cmd->msg.hdr.size);
-		if (*total_len != cmd->msg.hdr.size) {
-			FI_WARN(&smr_prov, FI_LOG_EP_CTRL,
-				"mmap iov copy out error\n");
-			ret = (*total_len < 0)
-				? ((int) *total_len) : -FI_ETRUNC;
-			goto munmap;
-		}
 	}
 
-munmap:
+	if (hmem_copy_ret < 0) {
+		FI_WARN(&smr_prov, FI_LOG_EP_CTRL,
+			"mmap copy iov failed with code %d\n",
+			(int)(-hmem_copy_ret));
+		ret = hmem_copy_ret;
+	} else if (hmem_copy_ret != cmd->msg.hdr.size) {
+		FI_WARN(&smr_prov, FI_LOG_EP_CTRL,
+			"mmap copy iov truncated\n");
+		ret = -FI_ETRUNC;
+	}
+
+	*total_len = hmem_copy_ret;
+
 	munmap(mapped_ptr, cmd->msg.hdr.size);
 unlink_close:
 	shm_unlink(shm_name);
@@ -441,6 +473,7 @@ static int smr_progress_ipc(struct smr_cmd *cmd, enum fi_hmem_iface iface,
 	uint64_t ipc_device;
 	int64_t id;
 	int ret, fd, ipc_fd;
+	ssize_t hmem_copy_ret;
 
 	peer_smr = smr_peer_region(ep->region, cmd->msg.hdr.id);
 	resp = smr_get_ptr(peer_smr, cmd->msg.hdr.src_data);
@@ -468,13 +501,13 @@ static int smr_progress_ipc(struct smr_cmd *cmd, enum fi_hmem_iface iface,
 		ptr = (char *) ptr + (uintptr_t) cmd->msg.data.ipc_info.offset;
 
 	if (cmd->msg.hdr.op == ofi_op_read_req) {
-		*total_len = ofi_copy_from_hmem_iov(ptr, cmd->msg.hdr.size,
-						    cmd->msg.data.ipc_info.iface,
-						    device, iov, iov_count, 0);
+		hmem_copy_ret = ofi_copy_from_hmem_iov(ptr, cmd->msg.hdr.size,
+						       cmd->msg.data.ipc_info.iface,
+						       device, iov, iov_count, 0);
 	} else {
-		*total_len = ofi_copy_to_hmem_iov(cmd->msg.data.ipc_info.iface,
-						  device, iov, iov_count, 0, ptr,
-						  cmd->msg.hdr.size);
+		hmem_copy_ret = ofi_copy_to_hmem_iov(cmd->msg.data.ipc_info.iface,
+						     device, iov, iov_count, 0,
+						     ptr, cmd->msg.hdr.size);
 	}
 
 	if (cmd->msg.data.ipc_info.iface == FI_HMEM_ZE)
@@ -483,8 +516,13 @@ static int smr_progress_ipc(struct smr_cmd *cmd, enum fi_hmem_iface iface,
 	/* Truncation error takes precedence over close_handle error */
 	ret = ofi_hmem_close_handle(cmd->msg.data.ipc_info.iface, base);
 
-	if (*total_len != cmd->msg.hdr.size)
-		ret = (*total_len < 0) ? ((int) *total_len) : -FI_ETRUNC;
+	if (hmem_copy_ret < 0) {
+		ret = hmem_copy_ret;
+	} else if (hmem_copy_ret != cmd->msg.hdr.size) {
+		ret = -FI_ETRUNC;
+	}
+
+	*total_len = hmem_copy_ret;
 
 out:
 	//Status must be set last (signals peer: op done, valid resp entry)

